### PR TITLE
Feature/users basic api

### PR DESCRIPTION
**Pull request by andreastihor**

### CODE REVIEW

File: api/services/users.js [hunk 1]
```diff
@@ -1,9 +1,10 @@
-const User = require('../../models').user
+const user = require('../../models').user
 
-module.exports.login(user) => {
-  return User.findOne({ where: { email: user.email } })
+module.exports.login = async (email, password) => {
```
Comment (vincentkdeli) [inline api/services/users.js]: no need to make it as async function if you don't use it asynchronously

File: api/services/users.js [hunk 2]
```diff
@@ -1,9 +1,10 @@
...
+  return user.findOne({ where: { email } })
     .then(userx => {
-      if (userx.password == user.password) {
+      let hashedPassword = Buffer.from(password).toString('base64')
```
Comment (vincentkdeli) [inline api/services/users.js]: move this before one line before return

File: api/services/users.js [hunk 3]
```diff
@@ -1,9 +1,10 @@
...
+      if (userx.password == hashedPassword) {
```
Comment (vincentkdeli) [inline api/services/users.js]: isn't it better to use '===' instead of '==' to check type as well?

File: api/controllers/report.js [hunk 4]
```diff
@@ -138,3 +235,224 @@ module.exports.deleteProduct = async (req, res) => {
...
+    .then(() => {
+      return res.response("Succeed in adding new Seasonal!")
+    })
+    .catch((err) => {
+      console.error(err)
+      return res.response("Error in adding new Seasonal")
+    })
+}
+
+//Addson
```
Comment (vincentkdeli) [inline api/controllers/report.js]: separate to single addson controller file, the file is getting too big

File: api/controllers/report.js [hunk 5]
```diff
@@ -138,3 +235,224 @@ module.exports.deleteProduct = async (req, res) => {
...
+    .then(() => {
+      return res.response("Succeed in adding new Category!")
+    })
+    .catch((err) => {
+      console.error(err)
+      return res.response("Error in adding new Category")
+    })
+}
+
+//Seasonal
```
Comment (vincentkdeli) [inline api/controllers/report.js]: separate to single seasonal controller file, the file is getting too big

File: api/controllers/report.js [hunk 6]
```diff
@@ -138,3 +235,224 @@ module.exports.deleteProduct = async (req, res) => {
     })
 
 }
+
+//Category
```
Comment (vincentkdeli) [inline api/controllers/report.js]: separate to single category controller file, the file is getting too big

File: api/controllers/report.js [hunk 7]
```diff
@@ -22,10 +23,97 @@ module.exports.createUser = async (req, res) => {
...
+      return res.response("Success Changing User Role")
+    })
+    .catch((err) => {
+      console.log(err);
+      return res.response("Failed in Changing User Role")
+    })
+}
+
+
 //Product
```
Comment (vincentkdeli) [inline api/controllers/report.js]: separate to single product controller file, the file is getting too big

File: api/controllers/report.js [hunk 8]
```diff
@@ -1,4 +1,5 @@
-const { user, product, category, seasonal } = require('../../models')
+const { user, product, category, seasonal, adds_on } = require('../../models')
+const { login } = require('../services/users')
 
 //User
```
Comment (vincentkdeli) [inline api/controllers/report.js]: separate to single user controller file, the file is getting too big

File: api/services/users.js (hunk 1 above)
Comment (andreastihor) [reply api/services/users.js]: findOne is asynchronous so it need to be async right?

File: api/controllers/adds_on.js [hunk 9]
```diff
@@ -0,0 +1,79 @@
+const { adds_on } = require('../../models/')
+//Addson
+module.exports.getAllAddson = async (req, res) => {
+  const tempAddson = await adds_on.findAll().
+    catch((err) => {
```
Comment (vincentkdeli) [inline api/controllers/adds_on.js]: lambda periods shouldn't be on separate lines
`const tempAddson = await adds_on.findAll()
  .catch((err) => {`

File: api/controllers/adds_on.js [hunk 10]
```diff
@@ -0,0 +1,79 @@
...
+  const { id } = req.payload
+  const tempAddson = await queryOneAddson(id)
+  if (!tempAddson) {
+    return res.response("Cannot Find Adds on, Sorry")
+  }
+  return tempAddson.update({
+    deletedAt: new Date()
+  })
+    .then(() => {
+      return res.response("Succed in delete Addson")
```
Comment (vincentkdeli) [inline api/controllers/adds_on.js]: typo

File: api/controllers/product.js [hunk 11]
```diff
@@ -69,12 +37,22 @@ async function queryAllProduct() {
...
   });
-  return tempProduct
+  return res.response(tempProduct)
 }
 
 module.exports.getAllProduct = async (req, res) => {
-  const product = await queryAllProduct()
-  return res.response(product)
+  const tempProduct = await product.findAll({
+    include: [{
```
Comment (vincentkdeli) [inline api/controllers/product.js]: please mind indentations, more readable
`include: [`
`  { model: category },`
`  { model: seasonal }`
`]

File: api/controllers/product.js [hunk 12]
```diff
@@ -69,12 +37,22 @@ async function queryAllProduct() {
...
+      model: category
+    }, {
+      model: seasonal
+    }
+    ]
```
Comment (vincentkdeli) [inline api/controllers/product.js]: please mind indentations, more readable
```suggestion
    include: [
      { model: category },
      { model: seasonal }
    ]
```

File: api/controllers/product.js [hunk 13]
```diff
@@ -107,6 +82,9 @@ module.exports.getProductDetail = async (req, res) => {
 module.exports.updateProduct = async (req, res) => {
   const { id, name, quantity, category_id, size, seasonal_id, image, price } = req.payload
   const tempProduct = await queryOneProduct(id)
+  if (!tempProduct) {
+    return res.response("Cannot Find Product, Sorry")
+  }
```
Comment (vincentkdeli) [inline api/controllers/product.js]: you can actually ignore the curly braces if it only contains one line, less line of code is better (imo)

File: api/controllers/user.js [hunk 14]
```diff
@@ -0,0 +1,113 @@
...
+    return res.response("Cannot find User Sorry")
+  }
+  const hashedPassword = Buffer.from(password).toString('base64')
+  return temp.update({
+    name,
+    address,
+    phone_number,
+    city,
+    email,
+    password: hashedPassword,
```
Comment (vincentkdeli) [inline api/controllers/user.js]: for attributes that still has key value pair (not keys only), should be added last on the body

File: api/controllers/adds_on.js [hunk 15]
```diff
@@ -0,0 +1,79 @@
...
+      throw new Error(err)
+    });
+  return res.response(tempAddson)
+}
+
+async function queryOneAddson(id) {
+  const tempAddson = await adds_on.findOne({
+    where: {
+      id
+    },
```
Comment (vincentkdeli) [inline api/controllers/adds_on.js]: if the body contains only one line, make it in single line
`where: { id }`

File: api/controllers/adds_on.js [hunk 16]
```diff
@@ -0,0 +1,79 @@
...
+    catch((err) => {
+      console.error(err)
+      throw new Error(err)
+    });
+  return res.response(tempAddson)
+}
+
+async function queryOneAddson(id) {
+  const tempAddson = await adds_on.findOne({
+    where: {
```
Comment (vincentkdeli) [inline api/controllers/adds_on.js]: if the body contains only one line, make it in single line
```suggestion
    where: { id }
```

File: api/controllers/category.js [hunk 17]
```diff
@@ -0,0 +1,72 @@
...
+
+module.exports.updateCategory = async (req, res) => {
+  const { id, name } = req.payload
+  const tempCategory = await queryOneCategory(id)
+  if (!tempCategory) {
+    return res.response("Cannot Find Category, Sorry")
+  }
+  return tempCategory.update({
+    name
+  }).then(() => {
```
Comment (vincentkdeli) [inline api/controllers/category.js]: you usually start lambda on new lines, please make it consistent for this one also

File: api/controllers/user.js [hunk 18]
```diff
@@ -0,0 +1,113 @@
...
+      console.error(err)
+      return res.response("failed in Deleting User")
+    })
+
+}
+
+module.exports.login = async (req, res) => {
+  const { email, password } = req.payload
+  const user = await login(email, password)
+  if (typeof (user) == "string") {
```
Comment (vincentkdeli) [inline api/controllers/user.js]: better to type check use === insead of ==

File: api/controllers/user.js [hunk 19]
```diff
@@ -0,0 +1,113 @@
...
+  return res.response({
+    status: "succeed",
+    user
+  })
+  //discuss further with deli
+}
+
+module.exports.changeRole = async (req, res) => {
+  const { id, role } = req.payload
+  const tempUser = await user.findOne({ where: { id } })
```
Comment (vincentkdeli) [inline api/controllers/user.js]:
```suggestion
  const tempUser = await user.findOne({
    where: { id }
  })
```

File: api/services/users.js [hunk 20]
```diff
@@ -1,10 +1,11 @@
...
-module.exports.login(user) => {
-  return User.findOne({ where: { email: user.email } })
-    .then(userx => {
-      if (userx.password == user.password) {
-        return userx
+const { user } = require('../../models')
+//try to firebase
+module.exports.login = (email, password) => {
+  let hashedPassword = Buffer.from(password).toString('base64')
+  return user.findOne({ where: { email } })
```
Comment (vincentkdeli) [inline api/services/users.js]:
```suggestion
  return user.findOne({
    where: { email }
  })
```

File: api/controllers/user.js [hunk 21]
```diff
@@ -0,0 +1,113 @@
...
+    address,
+    phone_number,
+    postal_code,
+    city,
+    email,
+    password: hashedPassword,
+    role: 'USER',
+    createdAt: new Date(),
+    updatedAt: new Date(),
+  }).then(() => {
```
Comment (vincentkdeli) [inline api/controllers/user.js]: you usually start lambda on new lines, please make it consistent for this one also

File: api/controllers/user.js [hunk 22]
```diff
@@ -0,0 +1,113 @@
...
+    console.error(err)
+    return res.response("Failed to insert user")
+  })
+}
+
+async function queryOneUser(id) {
+  const tempUser = await user.findOne({
+    where: {
+      id
+    },
```
Comment (vincentkdeli) [inline api/controllers/user.js]: `where: { id }`

File: api/controllers/adds_on.js [hunk 23]
```diff
@@ -0,0 +1,79 @@
...
+  if (!tempAddson) {
+    return res.response("Cannot Find Adds on, Sorry")
+  }
+  return tempAddson.update({
+    name,
+    image,
+    max_size,
+    min_size,
+    seasonal_id
+  }).then(() => {
```
Comment (vincentkdeli) [inline api/controllers/adds_on.js]: you usually start lambda on new lines, please make it consistent for this one also

File: api/controllers/category.js [hunk 24]
```diff
@@ -0,0 +1,72 @@
+const { category } = require('../../models/')
+
+//Category
+module.exports.getAllCategory = async (req, res) => {
+  const tempCategory = await category.findAll().
```
Comment (vincentkdeli) [inline api/controllers/category.js]: lambda periods shouldn't be on separate lines
```suggestion
  const tempCategory = await category.findAll()
    .catch((err) => {
```

File: api/controllers/category.js [hunk 25]
```diff
@@ -0,0 +1,72 @@
...
+    catch((err) => {
+      console.error(err)
+      throw new Error(err)
+    });
+  return res.response(tempCategory)
+}
+
+async function queryOneCategory(id) {
+  const tempCategory = await category.findOne({
+    where: {
```
Comment (vincentkdeli) [inline api/controllers/category.js]:
```suggestion
    where: { id }
```

File: api/controllers/seasonal.js [hunk 26]
```diff
@@ -0,0 +1,71 @@
...
+      throw new Error(err)
+    });
+  return res.response(tempSeasonal)
+}
+
+async function queryOneSeasonal(id) {
+  const tempSeasonal = await category.findOne({
+    where: {
+      id
+    },
```
Comment (vincentkdeli) [inline api/controllers/seasonal.js]: `where: { id }`

File: api/controllers/seasonal.js [hunk 27]
```diff
@@ -0,0 +1,71 @@
...
+
+module.exports.updateSeasonal = async (req, res) => {
+  const { id, name } = req.payload
+  const tempSeasonal = await queryOneSeasonal(id)
+  if (!tempSeasonal) {
+    return res.response("Cannot Find Seasonal, Sorry")
+  }
+  return tempSeasonal.update({
+    name
+  }).then(() => {
```
Comment (vincentkdeli) [inline api/controllers/seasonal.js]: you usually start lambda on new lines, please make it consistent for this one also

File: api/controllers/user.js (hunk 19 above)
Comment (andreastihor) [reply api/controllers/user.js]: i put this all on 1 line already

File: api/services/users.js (hunk 20 above)
Comment (andreastihor) [reply api/services/users.js]: same as before

**Pull request merged**